src/main/cpp/src/datetime_parser.cu

```diff
@@ -0,0 +1,453 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <map>
+#include <numeric>
+#include <vector>
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/utilities/vector_factories.hpp>
+#include <cudf/strings/convert/convert_datetime.hpp>
+#include <cudf/strings/detail/converters.hpp>
+#include <cudf/strings/detail/strings_children.cuh>
+#include <cudf/strings/detail/utilities.cuh>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/span.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/wrappers/timestamps.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+#include <thrust/execution_policy.h>
+#include <thrust/for_each.h>
+#include <thrust/functional.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/logical.h>
+#include <thrust/optional.h>
+#include <thrust/pair.h>
+#include <thrust/transform.h>
+
+#include "datetime_parser.hpp"
+
+namespace {
+
+using timestamp_components = spark_rapids_jni::timestamp_components;
+
+/**
+ * Get the timestamp from epoch from a local date time in a specific time zone.
+ * Note: local date time may be overlap or gap, refer to `ZonedDateTime.of`
+ *
+ */
+__device__ cudf::timestamp_us
+create_timestamp_from_components_and_zone(timestamp_components local_timestamp_components,
+                                          cudf::string_view time_zone) {
+  // TODO: implements:
+  //   val localDateTime = LocalDateTime.of(localDate, localTime)
+  //   val zonedDateTime = ZonedDateTime.of(localDateTime, zoneId)
+  //   val instant = Instant.from(zonedDateTime) // main work
+  //   instantToMicros(instant)
```

Comment:
hint: us or ms given we return ```cudf::timestamp_us```?

Reply:
both us(cuDF use this) and microsecond(Spark use this) are pow(10, -6).
ms/millisecond are pow(10, -3)
In cuDF we use us.